### PR TITLE
Update to Quarkus 3.0.1.Final

### LINKED ISSUE #15878: CVE-2022-41946 - Information Exposure vulnerability in org.postgresql:postgresql
### Description

-   Package Manager: maven
-   Vulnerable module: org.postgresql:postgresql
-   Introduced through: org.keycloak:keycloak-quarkus-server-app@999-SNAPSHOT, org.keycloak:keycloak-quarkus-server@999-SNAPSHOT and others

### Detailed paths

-   _Introduced through_: org.keycloak:keycloak-quarkus-server-app@999-SNAPSHOT › org.keycloak:keycloak-quarkus-server@999-SNAPSHOT › io.quarkus:[quarkus-jdbc-postgresql@2.13.5.Final](mailto:quarkus-jdbc-postgresql@2.13.5.Final) › org.postgresql:postgresql@42.5.0

# Overview

[org.postgresql:postgresql](https://mvnrepository.com/artifact/org.postgresql/postgresql) is a Java JDBC 4.2 (JRE 8+) driver for PostgreSQL database.

Affected versions of this package are vulnerable to Information Exposure in the pgjdbc driver, which writes to the operating system's shared temp directory when the InputStream to either `PreparedStatement.setText(int, InputStream)` or `PreparedStatemet.setBytea(int, InputStream)` is larger than 2K. The temporary file is readable by other users. This is the default system behavior on Unix systems but not on MacOS.

**NOTE:** This vulnerability is only fixed for JDK 1.7. Systems using JDK 1.6 or below can work around the vulnerability by setting the environment variable `java.io.tmpdir` to a non-world-readable location.

# Remediation

Upgrade `org.postgresql:postgresql` to version 42.2.27, 42.3.8, 42.4.3, 42.5.1 or higher.

# References

-   [GitHub Commit](https://github.com/pgjdbc/pgjdbc/commit/34855cffb9079edf95cf77930f3e3b73f6a1b139)
-   [GitHub Commit](https://github.com/pgjdbc/pgjdbc/commit/9008dc9aade6dbfe4efafcd6872ebc55f4699cf5)
-   [GitHub Commit](https://github.com/pgjdbc/pgjdbc/commit/91b8c2ca600f82d2ec47c3fb54ca2036b000e320)
-   [GitHub Commit](https://github.com/pgjdbc/pgjdbc/commit/e73c6b6b53ff11bc0bd06c44130a880d3bf77c55)

### LINKED ISSUE #15879: [CVE-2022-4147] Access Restriction Bypass vulnerability in io.quarkus:quarkus-vertx-http
### Description

-   Package Manager: maven
-   Vulnerable module: io.quarkus:quarkus-vertx-http
-   Introduced through: org.keycloak:keycloak-quarkus-server-app@999-SNAPSHOT, org.keycloak:keycloak-quarkus-server@999-SNAPSHOT and others

### Detailed paths

-   _Introduced through_: org.keycloak:keycloak-quarkus-server-app@999-SNAPSHOT › org.keycloak:keycloak-quarkus-server@999-SNAPSHOT › io.quarkus:[quarkus-vertx-http@2.13.5.Final](mailto:quarkus-vertx-http@2.13.5.Final)

# Overview

[io.quarkus:quarkus-vertx-http](https://github.com/quarkusio/quarkus) is a Cloud Native, (Linux) Container First framework for writing Java applications.

Affected versions of this package are vulnerable to Access Restriction Bypass via the `CORS` filter which allows simple `GET` and `POST` requests with invalid Origin to proceed.

# Remediation

Upgrade `io.quarkus:quarkus-vertx-http` to version 2.14.2 or higher.

# References

-   [GitHub Commit](https://github.com/quarkusio/quarkus/commit/cc26704d3bed46ed1d5caee90f22894952a07182)
-   [RedHat Bugzilla Bug](https://bugzilla.redhat.com/show_bug.cgi?id=2148867)

### LINKED ISSUE #17475: CVE-2023-0482 - Creation of Temporary File With Insecure Permissions vulnerability in org.jboss.resteasy:resteasy-core
### Description

**Note**: The issue affects the Keycloak Operator and the Quarkus distribution

-   Package Manager: maven
-   Vulnerable module: org.jboss.resteasy:resteasy-core
-   Introduced through: org.keycloak:[keycloak-operator@999.0.0-SNAPSHOT](mailto:keycloak-operator@999.0.0-SNAPSHOT), io.quarkus:[quarkus-rest-client@2.13.7.Final](mailto:quarkus-rest-client@2.13.7.Final) and others

### Detailed paths

-   _Introduced through_: org.keycloak:[keycloak-operator@999.0.0-SNAPSHOT](mailto:keycloak-operator@999.0.0-SNAPSHOT) › io.quarkus:[quarkus-rest-client@2.13.7.Final](mailto:quarkus-rest-client@2.13.7.Final) › io.quarkus:[quarkus-resteasy-common@2.13.7.Final](mailto:quarkus-resteasy-common@2.13.7.Final) › org.jboss.resteasy:[resteasy-core@4.7.7.Final](mailto:resteasy-core@4.7.7.Final)

# Overview

[org.jboss.resteasy:resteasy-core](https://github.com/resteasy/Resteasy) is a JBoss.org project aimed at providing productivity frameworks for developing client and server RESTful applications and services in Java.

Affected versions of this package are vulnerable to Creation of Temporary File With Insecure Permissions such that the insecure `File.createTempFile()` which is used in the `DataSourceProvider`, `FileProvider` and `Mime4JWorkaround` classes creates temp files with insecure permissions that could be read by a local user.

# Remediation

Upgrade `org.jboss.resteasy:resteasy-core` to version 4.7.8.Final or higher.

# References

-   [GitHub Commit](https://github.com/resteasy/resteasy/commit/277ea6d15b793b53b170ae9d8f9aba83b82d33c8)
-   [GitHub Release](https://github.com/resteasy/resteasy/releases/tag/4.7.8.Final)
-   [RedHat Bugzilla Bug](https://bugzilla.redhat.com/show_bug.cgi?id=2166004)

### LINKED ISSUE #17476: CVE-2023-24815 - Directory Traversal vulnerability in io.vertx:vertx-web
### Description

**Note**: The issue affects the Keycloak Operator and the Quarkus distribution

-   Package Manager: maven
-   Vulnerable module: io.vertx:vertx-web
-   Introduced through: org.keycloak:[keycloak-operator@999.0.0-SNAPSHOT](mailto:keycloak-operator@999.0.0-SNAPSHOT), io.quarkiverse.operatorsdk:quarkus-operator-sdk@4.0.7 and others

### Detailed paths

-   _Introduced through_: org.keycloak:[keycloak-operator@999.0.0-SNAPSHOT](mailto:keycloak-operator@999.0.0-SNAPSHOT) › io.quarkiverse.operatorsdk:quarkus-operator-sdk@4.0.7 › io.quarkus:[quarkus-smallrye-health@2.13.7.Final](mailto:quarkus-smallrye-health@2.13.7.Final) › io.quarkus:[quarkus-vertx-http@2.13.7.Final](mailto:quarkus-vertx-http@2.13.7.Final) › io.vertx:vertx-web@4.3.4

# Overview

[io.vertx:vertx-web](https://github.com/vert-x3/vertx-web) is a HTTP web applications for Vert.x.

Affected versions of this package are vulnerable to Directory Traversal when running Vertx web applications that serve files using `StaticHandle`. If the mount point is a wildcard (`*`) then an attacker can exfiltrate any class path resource.

When computing the relative path to locate the resource, in case of wildcards, [the code](https://github.com/vert-x3/vertx-web/blob/62c0d66fa1c179ae6a4d57344631679a2b97e60f/vertx-web/src/main/java/io/vertx/ext/web/impl/Utils.java#L83) returns the user input (without validation) as the segment to lookup. Even though checks are performed to avoid escaping the sandbox, given that the input was not sanitized, `\` are not properly handled and an attacker can build a path that is valid within the classpath.

**Note:** This vulnerability can only be exploited on Windows Operating Systems and Windows File Systems

# Details

A Directory Traversal attack (also known as path traversal) aims to access files and directories that are stored outside the intended folder. By manipulating files with "dot-dot-slash (../)" sequences and its variations, or by using absolute file paths, it may be possible to access arbitrary files and directories stored on file system, including application source code, configuration, and other critical system files.

Directory Traversal vulnerabilities can be generally divided into two types:

-   **Information Disclosure**: Allows the attacker to gain information about the folder structure or read the contents of sensitive files on the system.

`st` is a module for serving static files on web pages, and contains a [vulnerability of this type](https://snyk.io/vuln/npm:st:20140206). In our example, we will serve files from the `public` route.

If an attacker requests the following URL from our server, it will in turn leak the sensitive private key of the root user.

```
curl http://localhost:8080/public/%2e%2e/%2e%2e/%2e%2e/%2e%2e/%2e%2e/root/.ssh/id_rsa
```

**Note** `%2e` is the URL encoded version of `.` (dot).

-   **Writing arbitrary files**: Allows the attacker to create or replace existing files. This type of vulnerability is also known as `Zip-Slip`.

One way to achieve this is by using a malicious `zip` archive that holds path traversal filenames. When each filename in the zip archive gets concatenated to the target extraction folder, without validation, the final path ends up outside of the target folder. If an executable or a configuration file is overwritten with a file containing malicious code, the problem can turn into an arbitrary code execution issue quite easily.

The following is an example of a `zip` archive with one benign file and one malicious file. Extracting the malicious file will result in traversing out of the target folder, ending up in `/root/.ssh/` overwriting the `authorized_keys` file:

```
2018-04-15 22:04:29 .....           19           19  good.txt
2018-04-15 22:04:42 .....           20           20  ../../../../../../root/.ssh/authorized_keys
```

# Remediation

Upgrade `io.vertx:vertx-web` to version 4.3.8 or higher.

# References

-   [GitHub Commit](https://github.com/vert-x3/vertx-web/commit/8a6b4ef8096a0b8714c1f10ec0133f4b42e48428)
-   [PoC](https://github.com/adrien-aubert-drovio/vertx-statichandler-windows-traversal-path-vulnerability)
-   [Vulnerable Code](https://github.com/vert-x3/vertx-web/blob/62c0d66fa1c179ae6a4d57344631679a2b97e60f/vertx-web/src/main/java/io/vertx/ext/web/impl/Utils.java#L83)

### LINKED ISSUE #17555: CVE-2023-0482- Creation of Temporary File With Insecure Permissions vulnerability in org.jboss.resteasy:resteasy-multipart-provider
**Note**: The issue affects the Keycloak Operator and the Quarkus distribution

### Description

-   Package Manager: maven
-   Vulnerable module: org.jboss.resteasy:resteasy-multipart-provider
-   Introduced through: org.keycloak:[keycloak-quarkus-server-app@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server-app@999.0.0-SNAPSHOT), org.keycloak:[keycloak-quarkus-server@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server@999.0.0-SNAPSHOT) and others

### Detailed paths

-   _Introduced through_: org.keycloak:[keycloak-quarkus-server-app@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server-app@999.0.0-SNAPSHOT) › org.keycloak:[keycloak-quarkus-server@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server@999.0.0-SNAPSHOT) › org.jboss.resteasy:[resteasy-multipart-provider@4.7.7.Final](mailto:resteasy-multipart-provider@4.7.7.Final)

# Overview

Affected versions of this package are vulnerable to Creation of Temporary File With Insecure Permissions such that the insecure `File.createTempFile()` which is used in the `DataSourceProvider`, `FileProvider` and `Mime4JWorkaround` classes creates temp files with insecure permissions that could be read by a local user.

# Remediation

Upgrade `org.jboss.resteasy:resteasy-multipart-provider` to version 4.7.8.Final or higher.

# References

-   [GitHub Commit](https://github.com/resteasy/resteasy/commit/277ea6d15b793b53b170ae9d8f9aba83b82d33c8)
-   [GitHub Release](https://github.com/resteasy/resteasy/releases/tag/4.7.8.Final)
-   [RedHat Bugzilla Bug](https://bugzilla.redhat.com/show_bug.cgi?id=2166004)

### LINKED ISSUE #17556: CVE-2022-45787 - Information Exposure vulnerability in org.apache.james:apache-mime4j-storage
### Description

-   Package Manager: maven
-   Vulnerable module: org.apache.james:apache-mime4j-storage
-   Introduced through: org.keycloak:[keycloak-quarkus-server-app@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server-app@999.0.0-SNAPSHOT), org.keycloak:[keycloak-quarkus-server@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server@999.0.0-SNAPSHOT) and others

### Detailed paths

-   _Introduced through_: org.keycloak:[keycloak-quarkus-server-app@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server-app@999.0.0-SNAPSHOT) › org.keycloak:[keycloak-quarkus-server@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server@999.0.0-SNAPSHOT) › org.jboss.resteasy:[resteasy-multipart-provider@4.7.7.Final](mailto:resteasy-multipart-provider@4.7.7.Final) › org.apache.james:apache-mime4j-storage@0.8.3

# Overview

Affected versions of this package are vulnerable to Information Exposure such that unproper permissions on the temporary files used by MIME4J TempFileStorageProvider may lead to information disclosure to other local users.

# Remediation

Upgrade `org.apache.james:apache-mime4j-storage` to version 0.8.9 or higher.

# References

-   [Apache Lists](https://lists.apache.org/thread/26s8p9stl1z261c4qw15bsq03tt7t0rj)
-   [GitHub Commit](https://github.com/apache/james-mime4j/commit/021eb79ba312fe5a7f99fa867ee5350aa5533069)

### LINKED ISSUE #17557: CVE-2023-0481 - Creation of Temporary File With Insecure Permissions vulnerability in io.quarkus.resteasy.reactive:resteasy-reactive-common
### Description

-   Package Manager: maven
-   Vulnerable module: io.quarkus.resteasy.reactive:resteasy-reactive-common
-   Introduced through: org.keycloak:[keycloak-quarkus-server-deployment@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server-deployment@999.0.0-SNAPSHOT), io.quarkus:[quarkus-hibernate-orm-deployment@2.13.7.Final](mailto:quarkus-hibernate-orm-deployment@2.13.7.Final) and others

### Detailed paths

-   _Introduced through_: org.keycloak:[keycloak-quarkus-server-deployment@999.0.0-SNAPSHOT](mailto:keycloak-quarkus-server-deployment@999.0.0-SNAPSHOT) › io.quarkus:[quarkus-hibernate-orm-deployment@2.13.7.Final](mailto:quarkus-hibernate-orm-deployment@2.13.7.Final) › io.quarkus:[quarkus-resteasy-reactive-server-spi-deployment@2.13.7.Final](mailto:quarkus-resteasy-reactive-server-spi-deployment@2.13.7.Final) › io.quarkus.resteasy.reactive:[resteasy-reactive-processor@2.13.7.Final](mailto:resteasy-reactive-processor@2.13.7.Final) › io.quarkus.resteasy.reactive:[resteasy-reactive@2.13.7.Final](mailto:resteasy-reactive@2.13.7.Final) › io.quarkus.resteasy.reactive:[resteasy-reactive-common@2.13.7.Final](mailto:resteasy-reactive-common@2.13.7.Final)

# Overview

[io.quarkus.resteasy.reactive:resteasy-reactive-common](https://search.maven.org/artifact/io.quarkus.resteasy.reactive/resteasy-reactive-common/3.0.0.Alpha3/jar) is a RESTEasy Reactive - Common Runtime

Affected versions of this package are vulnerable to Creation of Temporary File With Insecure Permissions via the usage of `File.createTempFile()` class in the `FileBodyHandler` serializer that causes temp files to be created with `-rw-r--r--` permissions.

# Remediation

Update `io.quarkus.resteasy.reactive:resteasy-reactive-common` to 2.16.1.

# References

-   [RedHat Bugzilla Bug](https://bugzilla.redhat.com/show_bug.cgi?id=2163533)

### LINKED ISSUE #19080: Vulnerable packages and or dependencies found in keycloak 21.0.1 quarkus distribution
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

dist/quarkus

### Describe the bug

anchore enterprise finds following packages and or dependecies in keyckloak:

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:sshd-common
https://github.com/advisories/GHSA-fhw8-8j55-vwgq

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:jackson-databind
https://github.com/advisories/GHSA-jjjh-jjxp-wpff

/opt/keycloak/lib/lib/main/org.wildfly.security.wildfly-elytron-1.20.1.Final.jar:wildfly-elytron
https://github.com/advisories/GHSA-jmj6-p2j9-68cp

/opt/keycloak/lib/lib/main/org.yaml.snakeyaml-1.33.jar:snakeyaml
https://github.com/advisories/GHSA-mjmj-j48q-9wg2

/opt/keycloak/lib/lib/main/io.vertx.vertx-web-4.3.4.jar:vertx-web
https://github.com/advisories/GHSA-53jx-vvf9-4x38

### Version

21.0.1

### Expected behavior

ssd-common is 2.7.0 should be at least 2.9.2
jackson databind is 2.13.3 should be at least 2.13.4.1
wildfly-elytron is 1.20.1,Final should be at least 1.20.3.Final
snakeyaml is 1.33 should be 2.0
vertx-web is 4.3.4 should be at least 4.3.8

### Actual behavior

exploitable vulnerabilities present
ssd-common is 2.7.0
jackson databind is 2.13.3 
wildfly-elytron is 1.20.1,Final 
snakeyaml is 1.33
vertx-web is 4.3.4 

### How to Reproduce?

build docker image with keycloak 21.0.1
run grype and or anchore enterprise

### Anything else?

if some of this vulnerabilities are not exploitable would be good to know why

### LINKED ISSUE #19286: CVE-2022-1471
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

account/api

### Describe the bug

since this info is publicly available reporintg it here:
keycloak latest version contains a vulnerable version of snakeyaml.:
       "VulnerabilityID": "CVE-2022-1471",
          "PkgName": "org.yaml:snakeyaml",
          "PkgPath": "opt/keycloak/lib/lib/main/org.yaml.snakeyaml-1.33.jar",
          "InstalledVersion": "1.33",
can you fix and or share impact in case of keycloak (since it depends on this lib).
https://github.com/google/security-research/security/advisories/GHSA-mjmj-j48q-9wg2

### Version

21.0.1

### Expected behavior

keycloak does not contain vulnerability CVE-2022-1471 or indicates impact of this cve on keycloak usage.

### Actual behavior

..

### How to Reproduce?

scan a built container image using trivy

### Anything else?

_No response_

Duplicates: https://github.com/keycloak/keycloak/issues/19080

### LINKED ISSUE #20010: Update to Quarkus 3.0.1.Final
### Description

Quarkus team released 3.0.1.Final. It is important to upgrade to fix a coupe of CVEs on transitive dependencies, plus a couple of bug fixes present on this release.

### PR DESCRIPTION
**Pull request by abstractj**
Closes #20010
Closes #19080
Closes #17476
Closes #19286
Closes #17557
Closes #17556
Closes #17555
Closes #17475
Closes #15879
Closes #15878

### CODE REVIEW

File: operator/pom.xml [hunk 1]
```diff
@@ -31,8 +31,8 @@
         <maven.compiler.target>11</maven.compiler.target>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-        <quarkus.operator.sdk.version>6.0.0.CR1</quarkus.operator.sdk.version>
-        <quarkus.version>3.0.0.CR1</quarkus.version>
+        <quarkus.operator.sdk.version>6.0.1</quarkus.operator.sdk.version>
+        <quarkus.version>3.0.1.Final</quarkus.version>
```
Comment (vmuzikar) [inline operator/pom.xml:35]: Operator SDK 6.0.1 [uses](https://github.com/quarkiverse/quarkus-operator-sdk/blob/6.0.1/pom.xml#L15) Quarkus 3.0.0.  This property is not overriding the Quarkus version that SDK uses. It is just for alignment of other dependencies so it needs to match whatever the SDK uses.
```suggestion
        <quarkus.version>3.0.0.Final</quarkus.version>
```

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@abstractj Thanks, added one comment.

**Review by mabartos — DISMISSED** [review]

**Review by vmuzikar — COMMENTED** [review]
@abstractj Thanks, LGTM. Could you please rebase?

**Review by vmuzikar — DISMISSED** [review]
@abstractj Thanks, LGTM (provided tests will pass).

**Comment by abstractj** [comment]
It seems that model tests are broken due to the updates to Infinispan. @keycloak/store I'd appreciate your input on it.

**Comment by ahus1** [comment]
@abstractj - I had a short look at it, and was able to reproduce the exception in my IDE. 

I had a look and as I'm not too familiar with the code I'm hesitant to suggest a fix. 

The exception is gone once the two Infinispan dependencies are downgraded, so I suggest to upgrade the Infinispan dependency in a separate issue/PR.

**Comment by abstractj** [comment]
Thank you @ahus1. Incorporated your suggestion to the changes.

**Review by ahus1 — DISMISSED** [review]
Approving the downgrade of Infinispan assuming the build is now green.

**Review by vmuzikar — APPROVED** [review]

**Pull request merged**